### PR TITLE
build & ci: Setup version naming scheme & auto release CI

### PR DESCRIPTION
**Pull request by Qubik65536**
### Description/描述

This Pull Request contains:

1. Changing the version naming scheme from `version+date` (e.g. `0.1+20230830`) to `version.patch-git_sha-stage` (e.g. `0.1.62+f752f4c-dev`), where as `patch`, `git_sha` and `stage` will be get automatically during builds.
	- This change have been made to make sure it's easier to correspond a GitHub Release to a release artifact file. I DON'T think this will have a huge (and bad) impact on the ease of distinguishing versions.
2. Setup a CI action to automatically build beta version and publish to (as pre-release):
	- GitHub Releases
	- Modrinth
	- CurseForge
3. Setup a CI action to automatically build stable version and publish to (as official & latest release):
	- GitHub Releases
	- Modrinth
	- CurseForge
4. Setup a CI action to automatically build dev build from Pull Request (and NOT publish).

### Related Issues/相关的 issue

 

* *n/a*

### Checklist/清单

- [x] Setup new version naming scheme
- [x] Setup PR dev release CI
- [x] Setup beta release CI
	- [x] Build
	- [x] Release to GitHub
	- [x] Release to Modrinth
	- [x] Release to CurseForge
- [x] Setup stable release CI 
	- [x] Build
	- [x] Release to GitHub
	- [x] Release to Modrinth
	- [x] Release to CurseForge

- [x] The issues this PR addresses are related to each other/此 PR 解决的 issue 是相关的
- [x] My changes are all related to the related issue above/我的修改都与上方的相关 issue 有关
- [x] My changes generate no new warnings/我的修改没有产生新的警告
- [x] My code builds and runs on my machine/我的代码在我的电脑上可以编译和运行
- [x] I documented my code/我已经为我的代码添加了注释/文档

### Screenshots

*n/a*

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -1,15 +1,50 @@
...
+        return latestCommits.size()
+    }
+
+    getVersionGit = { List paths ->
+        if (grgit == null) {
+            return "nogit"
+        }
+
+        List latestCommits = paths.isEmpty() ? grgit.log(maxCommits: 1) : grgit.log(paths: paths, maxCommits: 1)
+        return latestCommits.isEmpty() ? "uncommited" : "${latestCommits.get(0).id.substring(0, 7)}"
```
Comment (zly2006) [inline build.gradle:47]: why substring 0, 7?

Comment (Qubik65536) [reply build.gradle:47]: first to seventh char of the commit SHA
Index 0 inclusive and index 7 exclusive

Comment (zly2006) [reply build.gradle:47]: but why 7 hex digits? I prefer 8 as 4 bytes.

Comment (Qubik65536) [reply build.gradle:47]: > but why 7 hex digits? I prefer 8 as 4 bytes.

Because it is exactly what GitHub shows in the commit history

Comment (zly2006) [reply build.gradle:47]: well, ok

Comment (Qubik65536) [reply build.gradle:47]: And if you try:

> git rev-parse --short HEAD

By default, it will gives you a 7 character long SHA

> ad7e9c4

**Comment by zly2006** [comment]
I still want to show build time in version number so I can easily see if a bug report is using the latest version.

You can add build time to only beta or snapshot versions.

**Comment by Qubik65536** [comment]
> I still want to show build time in version number so I can easily see if a bug report is using the latest version.
> 
> You can add build time to only beta or snapshot versions.

I personally don't recommend to do that, especially when the built time does not directly correspond to a version, but if you really want it I might add a environment variable to control that.

**Comment by zly2006** [comment]
well, it's ok to use commit hash, but if you don't like build time github action id is also ok,  i just want to know if players are using the latest version. e.g. `x.x.x-aaa+build.<action id>`

**Comment by Qubik65536** [comment]
> well, it's ok to use commit hash, but if you don't like build time github action id is also ok, i just want to know if players are using the latest version. e.g. `x.x.x-aaa+build.<action id>`

Yeah I'll put the CI ID at the end if it's generated by GitHub CI. BTW, this is only on beta isn't it?

**Comment by zly2006** [comment]
ok, great job

**Comment by Qubik65536** [comment]
Ready for review @zly2006 
Recommended to wait till all actions checks.

**Pull request merged**